cpp/src/arrow/dataset/dataset_encryption_test.cc

```diff
@@ -0,0 +1,412 @@
+      int num_columns, const std::string_view& footer_id,
+      const std::string_view& footer_key) {
+    std::unordered_map<std::string, std::string> key_map;
+    // Add column keys
+    for (int i = 0; i < num_columns; i++) {
+      key_map.insert({std::string(column_ids[i]), std::string(column_keys[i])});
+    }
+    // Add footer key
+    key_map.insert({std::string(footer_id), std::string(footer_key)});
+
+    return key_map;
+  }
+
+  // A utility function to validate our files were written out */
+  void ValidateFilesExist(const std::shared_ptr<arrow::fs::FileSystem>& fs,
+                          const std::vector<std::string>& files) {
+    for (const auto& file_path : files) {
+      ASSERT_OK_AND_ASSIGN(auto result, fs->GetFileInfo(file_path));
+
+      ASSERT_EQ(result.type(), arrow::fs::FileType::File);
+    }
+  }
+
+  // Build a dummy table
+  std::shared_ptr<::arrow::dataset::InMemoryDataset> BuildTable() {
+    // Create an Arrow Table
+    auto schema = arrow::schema(
+        {arrow::field("a", arrow::int64()), arrow::field("b", arrow::int64()),
+         arrow::field("c", arrow::int64()), arrow::field("part", arrow::utf8())});
+
+    std::vector<std::shared_ptr<arrow::Array>> arrays(4);
+    arrow::NumericBuilder<arrow::Int64Type> builder;
+    ARROW_EXPECT_OK(builder.AppendValues({0, 1, 2, 3, 4, 5, 6, 7, 8, 9}));
+    ARROW_EXPECT_OK(builder.Finish(&arrays[0]));
+    builder.Reset();
+    ARROW_EXPECT_OK(builder.AppendValues({9, 8, 7, 6, 5, 4, 3, 2, 1, 0}));
+    ARROW_EXPECT_OK(builder.Finish(&arrays[1]));
+    builder.Reset();
+    ARROW_EXPECT_OK(builder.AppendValues({1, 2, 1, 2, 1, 2, 1, 2, 1, 2}));
+    ARROW_EXPECT_OK(builder.Finish(&arrays[2]));
+    arrow::StringBuilder string_builder;
+    ARROW_EXPECT_OK(
+        string_builder.AppendValues({"a", "b", "c", "d", "e", "f", "g", "h", "i", "j"}));
+    ARROW_EXPECT_OK(string_builder.Finish(&arrays[3]));
+
+    auto table = arrow::Table::Make(schema, arrays);
+
+    // Write it using Datasets
+    return std::make_shared<::arrow::dataset::InMemoryDataset>(table);
+  }
+
+  // Helper function to create crypto factory and setup
+  std::shared_ptr<::parquet::encryption::CryptoFactory> SetupCryptoFactory(
+      bool wrap_locally, const std::unordered_map<std::string, std::string>& key_list) {
+    auto crypto_factory = std::make_shared<::parquet::encryption::CryptoFactory>();
+
+    auto kms_client_factory =
+        std::make_shared<::parquet::encryption::TestOnlyInMemoryKmsClientFactory>(
+            wrap_locally, key_list);
+
+    crypto_factory->RegisterKmsClientFactory(kms_client_factory);
+
+    return crypto_factory;
+  }
+};
+// Write dataset to disk with encryption
+// The aim of this test is to demonstrate the process of writing a partitioned
+// Parquet file while applying distinct file encryption properties to each
+// file within the test. This is based on the selected columns.
```

Comment:
I just took another pass on this test and my understanding may be wrong. Only column `a` is encrypted and all partitioned files share the same encryption properties. How to understand `applying distinct file encryption properties to each file`?

Reply:
Thanks for confirming! It seems the test can be simplified a little bit. Do you feel good to authorize me to directly edit the file dataset_encryption_test.cc in your PR?

Reply:
Thanks, I have pushed some changes related to the cpp test.